### PR TITLE
fix: strip pasted code/diffs from user-prompt + tag tool-action metadata

### PR DESCRIPTION
**Pull request by 24601**
## Summary

`user-prompt.ts` and `post-tool-use.ts` upload content to the Honcho message stream that the server-side fact extractor reads as user-authored prose, even when it isn't. This produces durable misattribution facts in the user-peer's representation — e.g. `<user> changed buildOperatorPlan` from a pasted diff in an adversarial-review prompt the user actually typed:

> *"You are performing an adversarial code review. Review the provided git diff: ```diff +function buildOperatorPlan(...) ... ``` "*

The diff body rides `role: "user"` per the Anthropic Messages API, so the extractor treats `+`/`-` lines as the user's statements about their own code.

This PR strips pasted code/diffs/long-output from `userPeer` uploads and tags `aiPeer` tool-action uploads with explicit role metadata.

## The bug, concretely

Two sites:

1. **`src/hooks/user-prompt.ts`** queues the entire `prompt` string for upload to `userPeer` at session-end. Pastes in user prompts (review-this-diff, fix-this-stack-trace, etc.) carry `role: "user"` per the Messages API but are not user-authored prose. The extractor reads them as such.

2. **`src/hooks/post-tool-use.ts`** uploads `aiPeer.message("[Tool] Edited file: ...")` with metadata that has no role discriminator. `unified` observation mode reduces blast radius but does not eliminate it — directional deployments and any MCP tool that operates in directional scope still fold these into the user's representation as `<user> did X`.

The end result: durable cross-session memory contaminated with assistant-authored facts attributed to the user. We observed this in the wild after running with `directional` mode briefly — a sample of 50 conclusions on page 1 of `list_conclusions` showed ~10 misattributions (extrapolated estimate: ~12K of ~60K conclusions in a heavily-used workspace).

### `src/hooks/user-prompt.ts` — strip pastes before queuing

New `stripPastes()` helper redacts:

```ts
// 1. Markdown fenced code blocks
out = out.replace(/```[\s\S]*?```/g, () => "[code block removed]");

// 2. Runs of 3+ consecutive unified-diff lines
out = out.replace(/(?:^[+\-].*(?:\r?\n|$)){3,}/gm, () => "[diff removed]\n");

// 3. Lines >200 chars containing a slash-path
//    (long file-path-bearing output: stack traces, log dumps, JSON pastes)
```

If anything is redacted, the queued message is tagged with `metadata.type = "user_paste_not_speech"` so server-side extraction can filter cross-peer attribution.

The single-character `+`/`-` line case (e.g. `"Note: + means added, - means removed."`) is preserved — only runs of 3 or more consecutive prefixed lines are treated as a diff. Short path mentions in prose (`"Edit /Users/foo/file.ts please"`) are also preserved.

### `src/hooks/post-tool-use.ts` — explicit role metadata

```ts
metadata: {
  instance_id: instanceId || undefined,
  session_affinity: sessionName,
  type: "tool_action",
  subject: "ai_action_on_user_behalf",
}
```

### `src/cache.ts` — optional metadata on `QueuedMessage`

`QueuedMessage` gains an optional `metadata` field. `queueMessage()` accepts an optional 5th argument that is forwarded to `userPeer.message()` at session-end. Backwards compatible — existing callers and queued messages without metadata continue to work.

### `src/hooks/session-end.ts` — forward queued metadata

Spreads `msg.metadata` into the `userPeer.message()` metadata block at upload time so server-side extraction sees the role tags.

## Verification

```bash
$ bunx tsc --noEmit -p tsconfig.json
# clean

$ bun run scripts/test-strip-pastes.ts
  PASS  pure prose — must NOT redact
  PASS  fenced code block — must redact + strip code
  PASS  unified diff — must redact
  PASS  long path-bearing line — must redact
  PASS  short path-bearing line — must NOT redact
  PASS  adversarial-review pattern (the original bug) — must redact code, preserve prose
  PASS  empty prompt — must NOT redact
  PASS  only a single +/- line (not a runs-of-3+ diff) — must NOT redact
  8/8 passed
```

The repo currently has no test runner; `scripts/test-strip-pastes.ts` is a pragmatic regression check intended to be rerun manually before releases that touch `user-prompt.ts`. Happy to wire it into CI in a follow-up PR if desired (or to introduce `bun:test` for this and future tests).

## Backwards compatibility

- `queueMessage()` signature change is purely additive (5th arg is optional). Single in-tree caller updated.
- `QueuedMessage` interface change is purely additive (`metadata?` optional). Existing queue files on disk parse cleanly because the field is absent.
- No config schema changes. No new env vars.
- No behavior change for callers that don't paste content.

## Notes for reviewers

- The strip happens at the **upload boundary** in the plugin, not at extraction time on the server. This keeps the change scoped to the plugin and avoids server-side coordination, but it means downstream tools that read the message stream directly will see redacted markers (`[code block removed]`, `[diff removed]`, `[path/output removed]`) instead of the original paste content. That's intentional — the original content is not the user's prose and shouldn't be in the user-peer's record.
- The 200-char path-bearing-line heuristic is conservative. Real prose lines are rarely that long; tool-output blobs frequently are. If this triggers false positives in practice, lowering the threshold to 120 or adding a `metadata.contains_paste: true` softer marker would be a follow-up.
- `metadata.subject = "ai_action_on_user_behalf"` is a string convention. If you'd prefer a structured enum or a reserved namespace (`honcho.role` / `honcho.subject`), I'm happy to rename — just point me at the convention.
- This PR does not touch `saveMessages: true` policy. A deeper-cut alternative would be to stop uploading raw user prompts entirely and only upload extracted facts — but that's a much bigger change and out of scope here.

## Summary by CodeRabbit

* **New Features**
  * Messages can include optional metadata to preserve context and categorization across session uploads.
  * User prompts are preprocessed client-side to redact pasted code blocks, diffs, and long path-like lines before upload; redacted messages are marked accordingly.
  * Tool-generated actions are annotated so they are distinguished from user-authored content when forwarded.

* **Tests**
  * Added a regression test verifying preprocessing and redaction behavior across edge cases.

### CODE REVIEW

File: plugins/honcho/src/hooks/user-prompt.ts [hunk 1]
```diff
@@ -68,6 +68,48 @@ function shouldSkipContextRetrieval(prompt: string): boolean {
...
+  out = out.replace(/```[\s\S]*?```/g, () => {
+    redacted = true;
+    return "[code block removed]";
+  });
+
+  // 2. Runs of 3+ consecutive unified-diff lines (^[+-] at line start)
+  out = out.replace(/(?:^[+\-].*(?:\r?\n|$)){3,}/gm, () => {
+    redacted = true;
+    return "[diff removed]\n";
+  });
```
[the remark this answers is not in the document]
Comment (24601) [reply plugins/honcho/src/hooks/user-prompt.ts]: Addressed in `b0e14ff`. Summary:

- **Major (this thread)**: gated diff redaction on a real unified-diff anchor (`@@`, `--- a/`, or `+++ b/`). Once anchored, dropped the `{3,}` run requirement and now redact every diff metadata + content line individually with a single `[diff removed]` marker. Markdown bullet lists, `+/-` in prose, and short single-character mentions all pass through untouched.
- **Nitpick (`session-end.ts`)**: reordered the metadata spread so `...(msg.metadata || {})` comes BEFORE the explicit per-upload fields. Future callers that put `instance_id`/`session_affinity` in queued metadata can no longer override the per-upload values.
- **Nitpick (`scripts/test-strip-pastes.ts`)**: added the markdown-bullet case + a mixed `+/-` bullet variant + two raw unfenced unified-diff cases (with `--- a/`/`+++ b/` and with `@@` only). 12/12 pass.

The previous test that asserted bare `+/-` lines without an anchor must redact has been flipped to `expectRedacted: false` to match the new (correct) behavior.

File: plugins/honcho/scripts/test-strip-pastes.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  {
+    name: "empty prompt — must NOT redact",
+    input: "",
+    expectRedacted: false,
+  },
+  {
+    name: "only a single +/- line (not a runs-of-3+ diff) — must NOT redact",
+    input: "Note: + means added, - means removed.",
+    expectRedacted: false,
+  },
```
[the remark this answers is not in the document]
Comment (24601) [reply plugins/honcho/scripts/test-strip-pastes.ts:192]: Renamed in `5302961` per your suggestion: `"stray +/- chars in prose (no diff anchor) — must NOT redact"`. Behavior unchanged; case still validates the no-anchor → no-redact path.

**Comment by ajspig** [comment]
Thanks for filing!

This PR is now touching code that main has moved pretty far away from. 
- #73 replaced queueMessage()/session-end flushing with live uploads (save-user-message.ts), so the metadata-threading through cache.ts and session-end.ts has no upload path to attach to (#88 reduced session-end to logging and state cleanup only).
- #77 addressed the broader misattribution class directly: harness-injected content is skipped via isHarnessInjected(), git events are no longer attributed to the user peer, and [Session ended] markers are gone. #90 extends the harness-injected filtering to retrieval queries as well.
- The post-tool-use.ts tagging is moot by default: [Tool] uploads are now gated behind saveToolUse, which defaults to false.

The other scenario of pasting in code, is a more rare edge case. Feel free to make a PR for that specifically or something more configurable. Closing for staleness.
